### DISCUSSION TITLE
Trainer check

### DISCUSSION THREAD

**Discussion by gianscarpe** [Lightning-AI/pytorch-lightning#6029] [Idea pool]
```Add a Trainer.check(...) interface that checks all your subclassed methods for (programatically) correctness.```

I propose to add a check function inside LightningModule. This function uses inspect
library to check that each overriden function of the LightningModule subclass
respect the correct signature.

E.g., training_epoch_end, if defines, should have two inputs (self:
LightningModule and outputs: List) and should return None

The same check can be defined for LightiningCallbacks and
LightningDataModule. trainer.check should call each check function and raise
an specific exception (something like ProgrammaticallyError exception) if some
check fail.

Note: we can exploit object-orientation; the user can defined is own check
function for its own module

What do you think? Any suggestions are really appreciated! :)

https://github.com/PyTorchLightning/pytorch-lightning/issues/3244

**Comment by carmocca** [comment]
Hi @gianscarpe! Thanks for the suggestion

How does this differ from running a tool like `mypy` on the user code? Maybe the effort should be put into improving our typing throughout the codebase.

**Reply by gianscarpe** [reply]
Hi @carmocca, good point! I think static type checking is mandatory and typing should be a priority in a project like pytorch lightning. I think this feature is still interesting, as it allows the user to check its own code without requiring any extra libraries, but I got your point! The idea was first proposed by @williamFalcon, maybe he can answer here and gives his insights :)
